### PR TITLE
Added functionality to output daily leaderboard

### PR DESCRIPTION
**Pull request by low101043**
I have updated the code which means you can now output the leaderboard for that day.  (So !daily 6 will output the leaderboard for day 6 with the scores).

### CODE REVIEW

File: main.py [hunk 1]
```diff
@@ -148,5 +172,85 @@ async def keen(context):
...
+        i = 0
+        #This checks they have day data
+        for item in player:
+            print(i, item)
+            i += 1
+        if (i != 5):
+            pass
+        else:
+            if (day in player[4].keys()):
+                players_day.append(player)
```
Comment (Danyc0) [inline main.py:196]: You should be able to replace this with just (effectively) lines 195 and 196 so:
```
for player in players:
    if day in player[4]:
        players_day.append(player)
```

File: main.py [hunk 2]
```diff
@@ -148,5 +172,85 @@ async def keen(context):
...
+    else:
+        result = "```"
+
+        for place, player in enumerate(final_table_sorted[::-1]):
+            result += str(place+1) + ": " + str(player[0]) + " with " + str(player[1]) +" point(s)\n"
+        result += "```"
+
+    await context.send(result)
 
+bot.run(TOKEN)
```
Comment (Danyc0) [inline main.py:256]: Having a new line at the end of file is good practice

File: main.py [hunk 3]
```diff
@@ -148,5 +172,85 @@ async def keen(context):
...
+            
+
+    #players_day has all people who have finished one star for that day
+    first_star = []
+    second_star = []
+
+    # Adds all the players which has 
+    for player_day in players_day:
+        if ('1' in player_day[4][day].keys()):
+
```
Comment (Danyc0) [inline main.py:206]: Unnecessary whitespace

File: main.py [hunk 4]
```diff
@@ -148,5 +172,85 @@ async def keen(context):
...
+                players_day.append(player)
+            
+
+    #players_day has all people who have finished one star for that day
+    first_star = []
+    second_star = []
+
+    # Adds all the players which has 
+    for player_day in players_day:
+        if ('1' in player_day[4][day].keys()):
```
Comment (Danyc0) [inline main.py:205]: .keys() shouldn't be necessary

File: main.py [hunk 5]
```diff
@@ -148,5 +172,85 @@ async def keen(context):
...
+            
+            first_star.append([player_day[0], int(player_day[4][day]['1']['get_star_ts'])])
+
+        if ('2' in player_day[4][day].keys()):
```
Comment (Danyc0) [inline main.py:209]: .keys() shouldn't be necessary

File: main.py [hunk 6]
```diff
@@ -148,5 +172,85 @@ async def keen(context):
...
+
+    #players_day has all people who have finished one star for that day
+    first_star = []
+    second_star = []
+
+    # Adds all the players which has 
+    for player_day in players_day:
+        if ('1' in player_day[4][day].keys()):
+            
+            first_star.append([player_day[0], int(player_day[4][day]['1']['get_star_ts'])])
```
Comment (Danyc0) [inline main.py:207]: Should this be a tuple instead of a list? So:
`(player_day[0], int(player_day[4][day]['1']['get_star_ts']))`
instead of
`[player_day[0], int(player_day[4][day]['1']['get_star_ts'])]`

File: main.py [hunk 7]
```diff
@@ -148,5 +172,85 @@ async def keen(context):
...
+
+        if ('2' in player_day[4][day].keys()):
+            second_star.append([player_day[0], int(player_day[4][day]['2']['get_star_ts'])])
```
Comment (Danyc0) [inline main.py:210]: Should this be a tuple instead of a list? So:
`(player_day[0], int(player_day[4][day]['1']['get_star_ts']))`
instead of
`[player_day[0], int(player_day[4][day]['1']['get_star_ts'])]`

File: main.py [hunk 8]
```diff
@@ -148,5 +172,85 @@ async def keen(context):
...
+    
+    #Sorts the two lists on timestamps.(Used insertion as lists aren't too large and easy to change to order on correct data)
+    first_star_sorted = insertion_sort(first_star)
+    second_star_sorted = insertion_sort(second_star)
+
+    print(first_star_sorted)
+    print("\n\n")
+    print(second_star_sorted)
+    print("\n\n")
```
Comment (Danyc0) [inline main.py:219]: Could we reduce the number of print statements throughout? Just to keep the log as clean as possible

File: main.py [hunk 9]
```diff
@@ -148,5 +172,85 @@ async def keen(context):
...
+    final_table_unsorted = []
+
+    #Adds all the people from first list
+    for i, player in enumerate(first_star_sorted):#
+        final_table_unsorted.append([player[0], (len(players) - i)])
```
Comment (Danyc0) [inline main.py:224]: Should this be a tuple instead of a list? So:
`(player[0], (len(players) - i))`
instead of
`[player[0], (len(players) - i)]`

I think there are a few places throughout where lists are used, but I think tuples would be better, line 233 is another example

File: main.py [hunk 10]
```diff
@@ -148,5 +172,85 @@ async def keen(context):
...
+            final_table_unsorted[index] = [to_change[0], (to_change[1] + (len(second_star_sorted) - i))]
+    
+    #Sorts the table
+    final_table_sorted = insertion_sort(final_table_unsorted)
+
+    print(final_table_sorted)
+
+    #Outputs data
+    result = ""
+    if (final_table_sorted == []):
```
Comment (Danyc0) [inline main.py:245]: Could be `if not final_table_sorted:` which is a bit cleaner imo

File: main.py [hunk 11]
```diff
@@ -148,5 +172,85 @@ async def keen(context):
...
+        final_table_unsorted.append([player[0], (len(players) - i)])
+    
+    print(final_table_unsorted)
+    print("\n\n")
+
+    #Updates the list with all the people who got the second star and their score
+    for i, player in enumerate(second_star_sorted):
+        index = find_name(player[0], final_table_unsorted)
+        if (index == -1):
+            final_table_unsorted.append([player[0], (len(players) - i)])
```
Comment (Danyc0) [inline main.py:233]: Can't we guarantee that if you've got the second start, then you must also have got the first star? If so I think this check is unnecessary. If it's "just in case" and you really want to put the check in, then it would probably be better to output some sort of error message instead of silently continuing as if nothing is wrong.

**Review by Danyc0 — CHANGES_REQUESTED** [review]
I've made quite a lot of comments, but it's mostly minor things. Overall it looks good (I'll have to check the logic again in the next review)

File: main.py [hunk 12]
```diff
@@ -17,6 +17,29 @@
...
+        position = index
+        
+        while position > 0 and data_array[position - 1][1] > current_data[1]:
+            
+            data_array[position] = data_array[position - 1]
+            position = position - 1
+        
+        data_array[position] = current_data
+        
+    return data_array
```
[the remark this answers is not in the document]
Comment (low101043) [reply main.py:33]: Ummm.  Just realised the only reason I did it was because I needed to sort it on the timestamps but I can do that using .sort() and a helper function.  Thanks!

File: main.py (hunk 8 above)
Comment (low101043) [reply main.py:219]: Yeah will remove them now.  (Used a lot for testing)

File: main.py (hunk 11 above)
Comment (low101043) [reply main.py:233]: Cool thought that when I was commenting the code.  I'll remove it now.

File: main.py (hunk 10 above)
Comment (low101043) [reply main.py:245]: (Didn't know that worked like that)

File: main.py (hunk 9 above)
Comment (low101043) [reply main.py:224]: Should be using tuples throughout now

File: main.py (hunk 7 above)
Comment (low101043) [reply main.py:210]: As below

File: main.py [hunk 13]
```diff
@@ -17,6 +17,29 @@
...
+
+
+def find_name(to_find, to_search):
+
+    for i, data in enumerate(to_search):
+        if (data[0] == to_find):
+            return i
+
+    return -1
```
[the remark this answers is not in the document]
Comment (low101043) [reply main.py:42]: Will do that (Haven't used Python List Comprehensions for years so why I didn't feel confident using it)

[the remark this answers is not in the document]
Comment (low101043) [reply main.py:42]: Need to have the index so it can be removed from the list and then be re added

[the remark this answers is not in the document]
Comment (low101043) [reply main.py:42]: With the correct score

File: main.py [hunk 14]
```diff
@@ -148,5 +172,85 @@ async def keen(context):
...
+    if context.channel.name != 'advent-of-code':
+        return
+
+    players = get_players()
+
+    players_day = []
+
+    # Goes through all the players checking if they have data for that day and if they do adding to players_days
+    for player in players:
+        i = 0
```
[the remark this answers is not in the document]
Comment (low101043) [reply main.py:187]: Made the whole thing neater (and harder to break).

File: main.py [hunk 15]
```diff
@@ -148,5 +172,85 @@ async def keen(context):
...
+        #This checks they have day data
+        for item in player:
+            print(i, item)
+            i += 1
```
[the remark this answers is not in the document]
Comment (low101043) [reply main.py:191]: As above

File: main.py [hunk 16]
```diff
@@ -148,5 +172,85 @@ async def keen(context):
     await context.send(result)
 
 
-bot.run(TOKEN)
+@bot.command(name='daily', help='Will give the Daily Leaderboard for specified date')
```
[the remark this answers is not in the document]
Comment (low101043) [reply main.py:175]: Honestly that's SPaG and I don't really mind

**Comment by low101043** [comment]
I've updated the code to what you have suggested (bar finding the index).

File: main.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (low101043) [reply main.py:30]: Proper Explanation.  I use this function in line 210 to find the index of each item in second_star in final_list.  This then allows me to get the right data and update the value and insert it into the list in the correct index rewriting the value (which is being updated).  I haven't used .index() as I looking at the usernames in final_list which is the first element of each tuple and you cannot specify which element to check in a tuple.  Hope this makes it slightly clearer

File: main.py [hunk 17]
```diff
@@ -17,20 +17,8 @@
...
-        while position > 0 and data_array[position - 1][1] > current_data[1]:
-            
-            data_array[position] = data_array[position - 1]
-            position = position - 1
-        
-        data_array[position] = current_data
-        
-    return data_array
+def toSort(data):
+    return data[1]
```
Comment (Danyc0) [inline main.py:21]: This looks like the perfect opportunity for a simple lambda

File: main.py [hunk 18]
```diff
@@ -184,12 +172,8 @@ async def daily(context, day):
...
-        i = 0
         #This checks they have day data
-        for item in player:
-            print(i, item)
-            i += 1
-        if (i != 5):
+        if (len(player) <= 4):
             pass
         else:
             if (day in player[4].keys()):
```
Comment (Danyc0) [inline main.py]: I still don't really get this check, what is this actually checking?
If it's just checking to see if the player has a 'completion_day_level', I think all players will have that field, it might just be that they are empty, in which case it should work fine without the check

File: main.py (hunk 4 above)
Comment (Danyc0) [reply main.py:189]: As well as the .keys(), the parenthesis aren't necessary

File: main.py (hunk 5 above)
Comment (Danyc0) [reply main.py:193]: As well as the .keys(), the parenthesis aren't necessary

File: main.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (Danyc0) [reply main.py:30]: I think you should still be able to do:
`[i for i, item in enumerate(to_search) if item[0] == to_find][0]`
In words: Return a list of all the 'i's in "for i, item enumerate(to_search)" where item[0] == to_find, then take the first result (in the case where there is only one match it doesn't matter)

**Review by Danyc0 — CHANGES_REQUESTED** [review]
Much better. Still a few things though

File: main.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (low101043) [reply main.py:30]: Resolving this cos I'm doing what you suggested at bottom.

File: main.py (hunk 2 above)
Comment (low101043) [reply main.py:232]: Added

File: main.py (hunk 1 above)
Comment (low101043) [reply main.py]: Going to resolve as you've recommented down below on the latest update

File: main.py (hunk 4 above)
Comment (low101043) [reply main.py:189]: Ahh cool didn't know I didn't need the .keys()  (Also sorry for parenthesis been doing a lot more Java and C so it's a bit hard coded at the moment!)

File: main.py (hunk 6 above)
Comment (low101043) [reply main.py]: Think I've done that last time

File: main.py (hunk 5 above)
Comment (low101043) [reply main.py:193]: Sorted

File: main.py (hunk 17 above)
Comment (low101043) [reply main.py:21]: Never used Lambdas in Python but now added to parts which need it

File: main.py (hunk 18 above)
Comment (low101043) [reply main.py]: So I had been an idiot and when you made anonymous player stuff I forgot to add the completion_day_level to their data which led to this check.  All sorted now

**Comment by low101043** [comment]
Sorted all the stuff now!  Understood the list comprehension now and am using that.  Also added a length check at end (as you did for main leaderboard) just saying to tell you (It does not go over with the first day info at the moment so unlikely to ever be needed)

File: main.py [hunk 19]
```diff
@@ -148,5 +148,72 @@ async def keen(context):
     await context.send(result)
 
 
-bot.run(TOKEN)
+@bot.command(name='daily', help='Will give the daily leaderboard for specified data')
```
Comment (Danyc0) [inline main.py:151]: For the specified day?

File: main.py [hunk 20]
```diff
@@ -148,5 +148,72 @@ async def keen(context):
...
+        return
+
+    players = get_players()
+
+    players_day = []
+
+    # Goes through all the players checking if they have data for that day and if they do adding to players_days
+    for player in players:
+        if day in player[4]:
+            players_day.append(player)
```
Comment (Danyc0) [inline main.py:164]: This can be compressed into a nice list comprehension

File: main.py [hunk 21]
```diff
@@ -148,5 +148,72 @@ async def keen(context):
...
+            
+
+    #players_day has all people who have finished one star for that day
+    first_star = []
+    second_star = []
+
+    # Adds all the players which has 
+    for player_day in players_day:
+        if '1' in player_day[4][day]:
+
```
Comment (Danyc0) [inline main.py:174]: Unnecessary whitespace

File: main.py [hunk 22]
```diff
@@ -148,5 +148,72 @@ async def keen(context):
...
+    #players_day has all people who have finished one star for that day
+    first_star = []
+    second_star = []
+
+    # Adds all the players which has 
+    for player_day in players_day:
+        if '1' in player_day[4][day]:
+            
+            first_star.append((player_day[0], int(player_day[4][day]['1']['get_star_ts'])))
+
```
Comment (Danyc0) [inline main.py:176]: Unnecessary whitespace

File: main.py [hunk 23]
```diff
@@ -148,5 +148,72 @@ async def keen(context):
...
+    first_star.sort(key=lambda data : data[1])
+    second_star.sort(key=lambda data : data[1])
 
+    final_table = []
+
+    #Adds all the people from first list
+    for i, player in enumerate(first_star):#
+        final_table.append((player[0], (len(players) - i)))
+    
+
```
Comment (Danyc0) [inline main.py:190]: Unnecessary whitespace

File: main.py [hunk 24]
```diff
@@ -148,5 +148,72 @@ async def keen(context):
...
+    #Updates the list with all the people who got the second star and their score
+    for i, player in enumerate(second_star):
+        index = [i for i, item in enumerate(final_table) if item[0] == player[0]][0]
+        to_change = final_table[index]
+        final_table[index] = (to_change[0], (to_change[1] + (len(second_star) - i)))
+    
+    #Sorts the table
+    final_table.sort(reverse=True,key=lambda data : data[1])
+
```
Comment (Danyc0) [inline main.py:200]: Unnecessary whitespace

File: main.py [hunk 25]
```diff
@@ -148,5 +148,72 @@ async def keen(context):
...
+    print(final_table)
```
Comment (Danyc0) [inline main.py:201]: This can either be deleted or commented out if you really want to keep it

File: main.py [hunk 26]
```diff
@@ -148,5 +148,72 @@ async def keen(context):
...
+
+    #Outputs data
+    result = ""
+    if not final_table:
+        result = "```No Scores for this day yet```"
+    else:
+        result = "```"
+
+        for place, player in enumerate(final_table):
+            result += str(place+1) + ": " + str(player[0]) + " with " + str(player[1]) +" point(s)\n"
```
Comment (Danyc0) [inline main.py:211]: It would be nice if this format could:
A. Use str.format() instead of concatenating a bunch of strings together, and
B. Use the same output format as is used for the other commands, the format string is available on line 18 and there are a couple of examples where you can see how it's used in the code

File: main.py [hunk 27]
```diff
@@ -148,5 +148,72 @@ async def keen(context):
...
+        result += "```"
+
+    if len(result) > 2000:
+        result = "```Tell Dan that this needs updating!```"
```
Comment (Danyc0) [inline main.py:215]: Maybe make this a bit more useful/similar to the other error messages, having my name in here doesn't really help if this were to be used in a different server

**Review by Danyc0 — CHANGES_REQUESTED** [review]
Does the points calculation work the same way as the official AoC one? (assuming it was running on single day)
Also could all comments start with a space and a capital letter.
Just a few tiny things, should only take you 5 minutes

File: main.py (hunk 19 above)
Comment (low101043) [reply main.py:151]: Yep.  Thanks

File: main.py (hunk 20 above)
Comment (low101043) [reply main.py:164]: Converted

File: main.py (hunk 25 above)
Comment (low101043) [reply main.py:201]: Done

File: main.py (hunk 27 above)
Comment (low101043) [reply main.py:215]: Changed it to output the top 40

File: main.py (hunk 26 above)
Comment (low101043) [reply main.py:211]: Sorted

**Comment by low101043** [comment]
Done all mentioned comments.

File: main.py [hunk 28]
```diff
@@ -148,32 +153,25 @@ async def keen(context):
...
-            
-
+    players_day = [player for player in players if day in player[4]]
+ 
     #players_day has all people who have finished one star for that day
     first_star = []
     second_star = []
 
-    # Adds all the players which has 
+    # Adds all the players which has stars into respective lists
```
Comment (Danyc0) [inline main.py:171]: "Adds all the players which have stars into the respective lists"

File: main.py [hunk 29]
```diff
@@ -17,6 +17,11 @@
 
 PLAYER_STR_FORMAT = '{rank:2}) {name:{name_pad}} ({points:{points_pad}}) {stars}* ({star_time})\n'
 
+def find_name(to_find, to_search):
+    return 
+
+    return -1
+
```
Comment (Danyc0) [inline main.py]: Where did this come from?

**Review by Danyc0 — CHANGES_REQUESTED** [review]
Looks like you might have accidentally pasted something random in.
One general thing that applies throughout, could all your comments start with a space and a capital letter?

File: main.py (hunk 29 above)
Comment (low101043) [reply main.py]: Accident!  Removed

File: main.py (hunk 28 above)
Comment (low101043) [reply main.py:171]: Sorted thanks

**Comment by low101043** [comment]
Sorted out the comments as well

File: main.py [hunk 30]
```diff
@@ -39,12 +38,13 @@ def get_players():
         players = [(data['members'][member]['name'],
                     data['members'][member]['local_score'],
                     data['members'][member]['stars'],
-                    int(data['members'][member]['last_star_ts'])) for member in data['members']]
+                    int(data['members'][member]['last_star_ts']),
+                    data['members'][member]['completion_day_level']) for member in data['members']] ##This is the data for all the days
```
[the remark this answers is not in the document]
Comment (Danyc0) [reply main.py:43]: You'll also have to add a type hint on line 152 to bring the day in as an integer

File: main.py [hunk 31]
```diff
@@ -148,5 +148,75 @@ async def keen(context):
...
+        max_name_len = len(max(final_table, key=lambda t: len(t[0]))[0])     
+        max_points_len = len(str((max(players, key=lambda t: len(str(t[1])))[1])))
+        result = "```"
+        for place, player in enumerate(final_table):
+            
+            result += PLAYER_STR_FORMAT.format(rank=place+1,
+                                        name=player[0], name_pad=max_name_len,
+                                        points=player[1], points_pad=max_points_len,
+                                        stars=player[3],
+                                        star_time=time.strftime('%H:%M %d/%m', time.localtime(player[2])))
```
Comment (Danyc0) [inline main.py:208]: Missing closing "```" for the result

File: main.py [hunk 32]
```diff
@@ -148,5 +148,75 @@ async def keen(context):
...
+
+    if len(result) > 2000:
+        result = "```"
+        for index in range(40):
```
Comment (Danyc0) [inline main.py:212]: This would catch fire if you had fewer than 40 players

**Review by Danyc0 — CHANGES_REQUESTED** [review]
Other than those I have mentioned in comments, there is one more bug I've spotted:

The number of stars seems to be 4* where it should be 2* (you can't get more than 2 stars in a given day)

File: main.py [hunk 33]
```diff
@@ -148,5 +148,75 @@ async def keen(context):
     await context.send(result)
 
 
-bot.run(TOKEN)
+@bot.command(name='daily', help='Will give the daily leaderboard for specified day')
+async def daily(context, day):
```
[the remark this answers is not in the document]
Comment (low101043) [reply main.py:153]: Only problem will be at 12-5 in the morning.  Will assume you are on the next day when we are still on the previous day.  At the moment line 175 catches all wrong inputs.  (I have added a check in case you haven't entered any day)

File: main.py (hunk 30 above)
[the remark this answers is not in the document]
Comment (low101043) [reply main.py:43]: That data is a dictionary not a string.

[the remark this answers is not in the document]
Comment (low101043) [reply main.py:43]: Unless I am missing something

File: main.py (hunk 31 above)
Comment (low101043) [reply main.py:208]: Done

File: main.py (hunk 32 above)
Comment (low101043) [reply main.py:212]: Cool I will resort out all of that

Comment (low101043) [reply main.py:212]: So changed it that it outputs every 10 people (Assuming that 10 people will have less than 2000 characters) and then will output the last set of people at the end

**Comment by low101043** [comment]
Have applied all the changes suggested apart from the one unresolved

**Comment by low101043** [comment]
Added new default value

File: main.py (hunk 30 above)
[the remark this answers is not in the document]
Comment (Danyc0) [reply main.py:43]: My objection here is that the keys for the dictionary should really be an integer (so 1 and 2 rather than '1' and '2'), which if they were, it would make life way simpler elsewhere in the code
In fact, really this shouldn't be a dict, it should just be a list (there's no point in having keys when the key is just the index+1)
Basically then, if you could maybe replace the dictionary with a list (possibly with something like this: `[star for star in data['members'][member]['completion_day_level']]`, if you do this, it could probably do with a comment) and then refactor the code below to use the integer indexes instead of the string keys

File: main.py [hunk 34]
```diff
@@ -39,12 +39,13 @@ def get_players():
         players = [(data['members'][member]['name'],
                     data['members'][member]['local_score'],
                     data['members'][member]['stars'],
-                    int(data['members'][member]['last_star_ts'])) for member in data['members']]
+                    int(data['members'][member]['last_star_ts']),
+                    data['members'][member]['completion_day_level']) for member in data['members']] ##This is the data for all the days
```
Comment (Danyc0) [inline main.py:43]: Could we avoid inline comments if possible? I think this comment would work well above the `players =` line

File: main.py [hunk 35]
```diff
@@ -148,5 +149,76 @@ async def keen(context):
...
+        max_name_len = len(max(final_table, key=lambda t: len(t[0]))[0])     
+        max_points_len = len(str((max(players, key=lambda t: len(str(t[1])))[1])))
+        result = "```"
+        for place, player in enumerate(final_table):
+           
+            result += PLAYER_STR_FORMAT.format(rank=place+1,
+                                        name=player[0], name_pad=max_name_len,
+                                        points=player[1], points_pad=max_points_len,
+                                        stars=player[3],
+                                        star_time=time.strftime('%H:%M %d/%m', time.localtime(player[2])))
```
Comment (Danyc0) [inline main.py:208]: Can this line up with the "rank=" part of the top line, see where this format  string is used elsewhere for examples

File: main.py [hunk 36]
```diff
@@ -148,5 +149,76 @@ async def keen(context):
...
+                result += "```"
+                await context.send(result)
+                result = "```"
+            
+
+    # This will output the rest of the list
+    result += "```" 
+
+    await context.send(result)
+
```
Comment (Danyc0) [inline main.py:221]: One too many blank lines

File: main.py [hunk 37]
```diff
@@ -148,5 +149,76 @@ async def keen(context):
...
+    result = ""
+    if not final_table:
+        result = "```No Scores for this day yet```"
+    else:
+            # Get string lengths for the format string
+        max_name_len = len(max(final_table, key=lambda t: len(t[0]))[0])     
+        max_points_len = len(str((max(players, key=lambda t: len(str(t[1])))[1])))
+        result = "```"
+        for place, player in enumerate(final_table):
+
```
Comment (Danyc0) [inline main.py:203]: Unnecessary whitespace

File: main.py [hunk 38]
```diff
@@ -148,5 +149,76 @@ async def keen(context):
...
+    
+    # Sorts the table
+    final_table.sort(reverse=True,key=lambda data : data[1])
+
+    # Outputs data
+    result = ""
+    if not final_table:
+        result = "```No Scores for this day yet```"
+    else:
+            # Get string lengths for the format string
```
Comment (Danyc0) [inline main.py:198]: Comment seems to be out of alignment

File: main.py [hunk 39]
```diff
@@ -148,5 +149,76 @@ async def keen(context):
     await context.send(result)
 
 
-bot.run(TOKEN)
+@bot.command(name='daily', help='Will give the daily leaderboard for specified day')
+async def daily(context, day : str = str(time.localtime(time.time())[7] - 335)):
```
Comment (Danyc0) [inline main.py:153]: `datetime.datetime.today().day` would be much cleaner

File: main.py (hunk 34 above)
Comment (Danyc0) [reply main.py:43]: And the missing space after the '#'

**Review by Danyc0 — CHANGES_REQUESTED** [review]
A bunch of these are just me being picky, but the big one is the ints instead of strings thing

File: main.py [hunk 40]
```diff
@@ -148,5 +149,76 @@ async def keen(context):
...
+        result = "```"
+        for place, player in enumerate(final_table):
+           
+            result += PLAYER_STR_FORMAT.format(rank=place+1,
+                                        name=player[0], name_pad=max_name_len,
+                                        points=player[1], points_pad=max_points_len,
+                                        stars=player[3],
+                                        star_time=time.strftime('%H:%M %d/%m', time.localtime(player[2])))
+                
+                # This will output every 10 people.
```
Comment (Danyc0) [inline main.py:210]: Comment out of alignment

Comment (Danyc0) [reply main.py:210]: Also, why 10? I think one message should support up to about 35-45 results

**Review by Danyc0 — CHANGES_REQUESTED** [review]
Couple of little things

File: main.py [hunk 41]
```diff
@@ -148,5 +149,76 @@ async def keen(context):
     await context.send(result)
 
 
-bot.run(TOKEN)
+@bot.command(name='daily', help='Will give the daily leaderboard for specified day')
+async def daily(context, day : str = str(time.localtime(time.time())[7] - 335)):
+    # Only respond if used in a channel called 'advent-of-code'
+    if context.channel.name != 'advent-of-code':
+        return
+
```
Comment (Danyc0) [inline main.py:157]: Could you add a quick debug print something along the lines of "Daily leaderboard requested for day: <the day>"?

File: main.py (hunk 35 above)
Comment (low101043) [reply main.py:208]: Done

File: main.py (hunk 36 above)
Comment (low101043) [reply main.py:221]: Done

File: main.py (hunk 37 above)
Comment (low101043) [reply main.py:203]: Done

File: main.py (hunk 38 above)
Comment (low101043) [reply main.py:198]: Done

File: main.py (hunk 41 above)
Comment (low101043) [reply main.py:157]: Done

File: main.py (hunk 40 above)
Comment (low101043) [reply main.py:210]: Chose 10 as I knew it would be within 2000.  Will up it to 40 (as I have seen !leaderboard 40 being used with no problem)

Comment (low101043) [reply main.py:210]: Done now

File: main.py (hunk 39 above)
Comment (low101043) [reply main.py:153]: Done.  Much nicer thanks

File: main.py (hunk 34 above)
Comment (low101043) [reply main.py:43]: I've taken it out you already have a comment above which explains whats happening

**Comment by low101043** [comment]
Updated the code from review.  Have messaged you via Discord about the problem with the dictionaries

File: main.py [hunk 42]
```diff
@@ -148,5 +149,76 @@ async def keen(context):
...
+            if place % 40 == 39:
+                result += "```"
+                await context.send(result)
+                result = "```"
+            
+
+    # This will output the rest of the list
+    result += "```" 
+    if result != "``````":
+
```
Comment (Danyc0) [inline main.py:220]: Holy whitespace batman

File: main.py [hunk 43]
```diff
@@ -148,5 +149,76 @@ async def keen(context):
...
+        max_points_len = len(str((max(players, key=lambda t: len(str(t[1])))[1])))
+        result = "```"
+        for place, player in enumerate(final_table):
+            result += PLAYER_STR_FORMAT.format(rank=place+1,
+                                               name=player[0], name_pad=max_name_len,
+                                               points=player[1], points_pad=max_points_len,
+                                               stars=player[3],
+                                               star_time=time.strftime('%H:%M %d/%m', time.localtime(player[2])))
+                
+            # This will output every 10 people.
```
Comment (Danyc0) [inline main.py:210]: Comment is out of date

**Review by Danyc0 — CHANGES_REQUESTED** [review]

File: main.py [hunk 44]
```diff
@@ -148,5 +149,76 @@ async def keen(context):
...
+    final_table.sort(reverse=True,key=lambda data : data[1])
+
+    # Outputs data
+    result = ""
+    if not final_table:
+        result = "```No Scores for this day yet"
+    else:
+        # Get string lengths for the format string
+        max_name_len = len(max(final_table, key=lambda t: len(t[0]))[0])     
+        max_points_len = len(str((max(players, key=lambda t: len(str(t[1])))[1])))
```
Comment (Danyc0) [inline main.py:201]: This should be `final_table` not `players`, this causes a bug where if the max total score is 3 digits long and the max daily score is only 2, there is an excess padding character on all the scores

**Review by Danyc0 — CHANGES_REQUESTED** [review]

File: main.py (hunk 30 above)
[the remark this answers is not in the document]
Comment (Danyc0) [reply main.py]: I'll change this in a future commit, closing for now

File: main.py [hunk 45]
```diff
@@ -148,5 +149,76 @@ async def keen(context):
     await context.send(result)
 
 
-bot.run(TOKEN)
+@bot.command(name='daily', help='Will give the daily leaderboard for specified day')
+async def daily(context, day : str = str(datetime.datetime.today().day)):
```
Comment (Danyc0) [inline main.py:153]: Replacing:
`str(datetime.datetime.today().day)`
with
`str(datetime.datetime.today() - datetime.timedelta(hours=5)).day`
should hopefully compensate for the time difference. The server time is in UTC, so this should convert it to the UTC-5 the challenges come out in. A comment to this effect would be useful too (on the line below this, i.e. the first line of the function)

File: main.py (hunk 42 above)
Comment (low101043) [reply main.py:220]: Oops!  Sorted

File: main.py (hunk 43 above)
Comment (low101043) [reply main.py:210]: Done

File: main.py (hunk 44 above)
Comment (low101043) [reply main.py:201]: Sorted

File: main.py (hunk 45 above)
Comment (low101043) [reply main.py:153]: Sorted if you can just check the comment for me

File: main.py [hunk 46]
```diff
@@ -148,5 +149,76 @@ async def keen(context):
     await context.send(result)
 
 
-bot.run(TOKEN)
+@bot.command(name='daily', help='Will give the daily leaderboard for specified day')
+async def daily(context, day : str = str((datetime.datetime.today() - datetime.timedelta(hours=5)).day)):
+    # The default input is whatever day's competition has just come out.  SO at 2AM will still show previous day's leaderboard
```
Comment (Danyc0) [inline main.py:154]: Typo "SO" -> "So"
Probably say "default day" instead of "default input"
Maybe say "challenge" instead of "competition"? I'd say a competition is a series of challenges.
I'd also use 4:59AM UTC as the example rather than 2AM

**Review by Danyc0 — CHANGES_REQUESTED** [review]
Looks really good, just update the comment

File: main.py [hunk 47]
```diff
@@ -148,5 +149,76 @@ async def keen(context):
...
+    second_star = []
+
+    # Adds all the players which has stars the into respective lists
+    for player_day in players_day:
+        if '1' in player_day[4][day]:
+            first_star.append((player_day[0], int(player_day[4][day]['1']['get_star_ts'])))
+        if '2' in player_day[4][day]:
+            second_star.append((player_day[0], int(player_day[4][day]['2']['get_star_ts'])))
+    
+    # Sorts the two lists on timestamps.(Used insertion as lists aren't too large and easy to change to order on correct data)
```
Comment (Danyc0) [inline main.py:176]: Comment is now out of date (you can just remove the full stop and the bit in brackets)

File: main.py [hunk 48]
```diff
@@ -148,5 +149,76 @@ async def keen(context):
...
+
+    # Adds all the people from first list
+    for i, player in enumerate(first_star):
+        final_table.append((player[0], (len(players) - i), player[1], 1))  
+    
+    # Updates the list with all the people who got the second star and their score
+    for i, player in enumerate(second_star):
+        index = [i for i, item in enumerate(final_table) if item[0] == player[0]][0]
+        to_change = final_table[index]
+        final_table[index] = (to_change[0], (to_change[1] + (len(second_star) - i)), player[1], 2)
```
Comment (Danyc0) [inline main.py:190]: I think `len(second_star)` should be `len(players)` because the score is based on your speed compared to all players, not compared to all players who got the star. Line 184 seems to have done the right thing for the first star

File: main.py (hunk 46 above)
Comment (low101043) [reply main.py:154]: Done

File: main.py (hunk 47 above)
Comment (low101043) [reply main.py:176]: Done

File: main.py (hunk 48 above)
Comment (low101043) [reply main.py:190]: Good spot thanks

**Comment by low101043** [comment]
Sorted out the comments and stuff

File: main.py [hunk 49]
```diff
@@ -148,5 +149,76 @@ async def keen(context):
     await context.send(result)
 
 
-bot.run(TOKEN)
+@bot.command(name='daily', help='Will give the daily leaderboard for specified day')
+async def daily(context, day : str = str((datetime.datetime.today() - datetime.timedelta(hours=5)).day)):
+    # The default day is whatever day's challenge has just come out.  So at 4.59AM UTC will still show previous day's leaderboard
```
Comment (Danyc0) [inline main.py:154]: There's two spaces at the start of the second sentence

Comment (low101043) [reply main.py:154]: Sorry normally do 2 spaces!

File: main.py [hunk 50]
```diff
@@ -148,5 +149,76 @@ async def keen(context):
     await context.send(result)
 
 
-bot.run(TOKEN)
+@bot.command(name='daily', help='Will give the daily leaderboard for specified day')
+async def daily(context, day : str = str((datetime.datetime.today() - datetime.timedelta(hours=5)).day)):
+    # The default day is whatever day's challenge has just come out. So at 4.59AM UTC will still show previous day's leaderboard
```
Comment (Danyc0) [inline main.py:154]: Untested fix for this giving the incorrect day:
```
async def daily(context, day : str = None):
    # The default day calculation cannot be in the function default value because the default value is evaluated when the program is started, not when the function is called
    if day is None:
        # The default day is whatever day's challenge has just come out. So at 4.59AM UTC will still show previous day's leaderboard
        str((datetime.datetime.today() - datetime.timedelta(hours=5)).day)
```
